tests/mappings/mapping_q_eulerian_14.cc

```diff
@@ -0,0 +1,395 @@
+          << "n_active_cells: " << triangulation.n_active_cells() << '\n'
+          << "n_levels: " << triangulation.n_levels() << std::endl;
+
+  FE_Q<dim>       fe(fe_degree);
+  DoFHandler<dim> dof_handler(triangulation);
+  dof_handler.distribute_dofs(fe);
+  dof_handler.distribute_mg_dofs();
+
+  // quadrature for MatrixFree, not related to the degree in the
+  // FE displacement field.
+  QGauss<1> quadrature_formula(n_q_points);
+
+
+  FESystem<dim>   fe_euler(FE_Q<dim>(euler_fe_degree), dim);
+  DoFHandler<dim> dof_handler_euler(triangulation);
+  dof_handler_euler.distribute_dofs(fe_euler);
+  dof_handler_euler.distribute_mg_dofs();
+
+  // IndexSets and constraints
+  const IndexSet &locally_owned_dofs_euler =
+    dof_handler_euler.locally_owned_dofs();
+  IndexSet locally_relevant_dofs_euler;
+  DoFTools::extract_locally_relevant_dofs(dof_handler_euler,
+                                          locally_relevant_dofs_euler);
+
+  const IndexSet &locally_owned_dofs = dof_handler.locally_owned_dofs();
+  IndexSet        locally_relevant_dofs;
+  DoFTools::extract_locally_relevant_dofs(dof_handler, locally_relevant_dofs);
+
+  // constraints:
+  AffineConstraints<double> constraints_euler;
+  constraints_euler.reinit(locally_relevant_dofs_euler);
+  DoFTools::make_hanging_node_constraints(dof_handler_euler, constraints_euler);
+  constraints_euler.close();
+
+  AffineConstraints<double> constraints;
+  constraints.reinit(locally_relevant_dofs);
+  DoFTools::make_hanging_node_constraints(dof_handler, constraints);
+  constraints_euler.close();
+
+  // MG constraints:
+  MGConstrainedDoFs mg_constrained_dofs_euler;
+  mg_constrained_dofs_euler.initialize(dof_handler_euler);
+
+  // Displacement vector
+  LinearAlgebra::distributed::Vector<NumberType> displacement;
+  displacement.reinit(locally_owned_dofs_euler,
+                      locally_relevant_dofs_euler,
+                      mpi_communicator);
+
+  VectorTools::project<dim,
+                       LinearAlgebra::distributed::Vector<NumberType>,
+                       dim>(dof_handler_euler,
+                            constraints_euler,
+                            QGauss<dim>(n_q_points),
+                            displacement_function,
+                            displacement);
+  displacement.update_ghost_values();
+
+  MGTransferMatrixFree<dim, LevelNumberType> mg_transfer_euler(
+    mg_constrained_dofs_euler);
+  mg_transfer_euler.build(dof_handler_euler);
+
+  // now the core of the test:
+  const unsigned int max_level =
+    dof_handler.get_triangulation().n_global_levels() - 1;
+  const unsigned int min_level = 0;
+  MGLevelObject<LinearAlgebra::distributed::Vector<LevelNumberType>>
+    displacement_level(min_level, max_level);
+  mg_transfer_euler.interpolate_to_mg(dof_handler_euler,
+                                      displacement_level,
+                                      displacement);
```

Comment:
The "bug" is here, and is fixed by adding this code before the `interpolate_to_mg` call:
```c++
  for (unsigned int level = 0; level <= max_level; ++level)
    {
      IndexSet relevant_mg_dofs;
      DoFTools::extract_locally_relevant_level_dofs(dof_handler_euler,
                                                    level,
                                                    relevant_mg_dofs);
      displacement_level[level].reinit(dof_handler_euler.locally_owned_mg_dofs(
                                         level),
                                       relevant_mg_dofs,
                                       mpi_communicator);
    }
```
This situation reveals an oversight in our multigrid classes, though: We somewhat naively assume that the `interpolate_to_mg` would need the same indices as the `restrict` and `prolongate` functions (implied by `copy_to_mg`). However, one might want to have more ghosts especially in the present context of `MappingQEulerian`, as we also want to be able to fill the information on all ghost cells. Or, as in the case here, it might even fail on just the locally owned cells when some of the cells get handled by another MPI process in the transfer setting.

In our CFD application code, we use a similar strategy, namely to copy to a vector to more ghosts after the transfer has been execute (in order not to rely on the fact that `copy_to_mg` preserves the ghost state on the level vectors), namely: https://github.com/exadg/exadg/blob/2eabd006f482001b22251632b3c97644a9918706/include/exadg/grid/mapping_dof_vector.h#L306-L322
The main question is whether we should always create vectors with all ghosts for the `interpolate_to_mg` functions, i.e., here https://github.com/dealii/dealii/blob/f37e0290dd67164a528f2c57e9e1586abc619b8a/include/deal.II/multigrid/mg_transfer_matrix_free.h#L524-L531 and https://github.com/dealii/dealii/blob/f37e0290dd67164a528f2c57e9e1586abc619b8a/include/deal.II/multigrid/mg_transfer_global_coarsening.h#L682-L684 would need to be adjusted accordingly. Or we need to document this more clearly. Opinions?

Reply:
> This situation reveals an oversight in our multigrid classes, though: We somewhat naively assume that the `interpolate_to_mg` would need the same indices as the `restrict` and `prolongate` functions

That makes sense and I can confirm that this fixes the issue. I was mislead because I used ``tests/mappings/mapping_q_eulerian_08|09`` as a basis for the code.

> In our CFD application code, we use a similar strategy, namely to copy to a vector to more ghosts after the transfer has been execute

Is there an advantage to doing it afterwards? Is this because you use the transfer for something else?

> The main question is whether we should always create vectors with all ghosts for the `interpolate_to_mg` functions

Do you think this would incur a significant cost? We probably only need this change when using ``MappingQEulerian``...